src/System.Net.Security/tests/FunctionalTests/ServerAllowNoEncryptionTest.cs

```diff
@@ -94,6 +95,6 @@ public async Task ServerAllowNoEncryption_ClientNoEncryption_ConnectWithNoEncryp
             }
         }
 
-        private static bool SupportsNullEncryption => TestConfiguration.SupportsNullEncryption;
+        private static bool SupportsNullEncryption => TestConfiguration.SupportsNullEncryption && !PlatformDetection.SupportsTls13;
     }
```

Comment:
That's a great question.  I thought that null encryption still works on platforms that support it generally speaking as long as the TLS version being used is less than TLS 1.3.  If so, this logic above is incorrect.

Reply:
I think this was alluded to in the comments.  The current state of affairs is "supports null encryption" on Linux means OpenSSL 1.1.0 or older.  "Supports TLS 1.3" is only Linux with OpenSSL 1.1.1 or higher.  So currently "supports TLS 1.3" implies "does not support null encryption".

Reply:
>I think this was alluded to in the comments.

I think the comments in the code need to be reworded to make this clearer.  Starting with @bartonjs comments would be good.

Reply:
Note that the TestConfiguration.SupportsNullEncryption  uses `openssl ciphers NULL` and checks for non-empty output.
With openssl 1.1.1 we get this:
```
openssl ciphers NULL
TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256:ECDHE-ECDSA-NULL-SHA:ECDHE-RSA-NULL-SHA:AECDH-NULL-SHA:NULL-SHA256:ECDHE-PSK-NULL-SHA384:ECDHE-PSK-NULL-SHA256:ECDHE-PSK-NULL-SHA:RSA-PSK-NULL-SHA384:RSA-PSK-NULL-SHA256:DHE-PSK-NULL-SHA384:DHE-PSK-NULL-SHA256:RSA-PSK-NULL-SHA:DHE-PSK-NULL-SHA:NULL-SHA:NULL-MD5:PSK-NULL-SHA384:PSK-NULL-SHA256:PSK-NULL-SHA
```
and
```
openssl ciphers -s NULL
TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256:ECDHE-ECDSA-NULL-SHA:ECDHE-RSA-NULL-SHA:AECDH-NULL-SHA:NULL-SHA256:NULL-SHA:NULL-MD5
````

I could update TestConfiguration.SupportsNullEncryption but I thought it may worth of keeping tests running so we have defined behavior. It seems like the behavior is now that we get back alert and throw Authentication exception instead of just getting IO error.

Reply:
FWIW, I'm a little confused.  I feel like when I did the OpenSSL 1.0/1.1 hybridization (which included testing with 1.1.1) all the tests passed; and that `SSL_CTX_set_security_level(ctx, 0)` should leave these tests working.  I see that there is still a call to `SSL_CTX_set_security_level(ctx, 0)` when AllowNoEncryption or NoEncryption is passed.

Is this that we changed my original "if you asked for no encryption I drop system default to TLS 1.0/1.1/1.2" to something else, and that we (might) want tests for confirming that these cases work when a TLS version is specified?

(Or maybe the tests weren't passing, or they're new since then, or my brain is full of fog.)

Reply:
I'm looking into it. This set of tests is different than the other. When I enable it on Ubuntu 18.10 I get this:
```
      System.Net.Security.Tests.ServerAllowNoEncryptionTest.ServerAllowNoEncryption_ClientNoEncryption_ConnectWithNoEncryption [STARTING]
      System.Net.Security.Tests.ServerAllowNoEncryptionTest.ServerAllowNoEncryption_ClientNoEncryption_ConnectWithNoEncryption [FAIL]
        System.Security.Authentication.AuthenticationException : Authentication failed, see inner exception.
        ---- Interop+OpenSsl+SslException : The 'NoEncryption' encryption policy is not supported by this installation of OpenSSL.
        Stack Trace:
          /home/furt/github/wfurt-corefx/src/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs(710,0): at System.Net.Security.SslStream.StartSendAuthResetSignal(ProtocolToken message, AsyncProtocolRequest asyncRequest, ExceptionDispatchInfo exception)
          /home/furt/github/wfurt-corefx/src/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs(537,0): at System.Net.Security.SslStream.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
          /home/furt/github/wfurt-corefx/src/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs(527,0): at System.Net.Security.SslStream.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)
          /home/furt/github/wfurt-corefx/src/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs(398,0): at System.Net.Security.SslStream.ForceAuthentication(Boolean receiveFirst, Byte[] buffer, AsyncProtocolRequest asyncRequest)
          /home/furt/github/wfurt-corefx/src/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs(309,0): at System.Net.Security.SslStream.ProcessAuthentication(LazyAsyncResult lazyResult)
          /home/furt/github/wfurt-corefx/src/System.Net.Security/src/System/Net/Security/SslStream.cs(233,0): at System.Net.Security.SslStream.BeginAuthenticateAsClient(SslClientAuthenticationOptions sslClientAuthenticationOptions, CancellationToken cancellationToken, AsyncCallback asyncCallback, Object asyncState)
          /home/furt/github/wfurt-corefx/src/System.Net.Security/src/System/Net/Security/SslStream.cs(222,0): at System.Net.Security.SslStream.BeginAuthenticateAsClient(String targetHost, X509CertificateCollection clientCertificates, SslProtocols enabledSslProtocols, Boolean checkCertificateRevocation, AsyncCallback asyncCallback, Object asyncState)
          /home/furt/github/wfurt-corefx/src/System.Net.Security/src/System/Net/Security/SslStream.cs(392,0): at System.Net.Security.SslStream.<>c.<AuthenticateAsClientAsync>b__64_1(String arg1, X509CertificateCollection arg2, SslProtocols arg3, AsyncCallback callback, Object state)
          /__w/1/s/src/System.Private.CoreLib/shared/System/Threading/Tasks/FutureFactory.cs(1152,0): at System.Threading.Tasks.TaskFactory`1.FromAsyncImpl[TArg1,TArg2,TArg3](Func`6 beginMethod, Func`2 endFunction, Action`1 endAction, TArg1 arg1, TArg2 arg2, TArg3 arg3, Object state, TaskCreationOptions creationOptions)
          /__w/1/s/src/System.Private.CoreLib/shared/System/Threading/Tasks/TaskFactory.cs(1093,0): at System.Threading.Tasks.TaskFactory.FromAsync[TArg1,TArg2,TArg3](Func`6 beginMethod, Action`1 endMethod, TArg1 arg1, TArg2 arg2, TArg3 arg3, Object state)
          /home/furt/github/wfurt-corefx/src/System.Net.Security/src/System/Net/Security/SslStream.cs(393,0): at System.Net.Security.SslStream.AuthenticateAsClientAsync(String targetHost, X509CertificateCollection clientCertificates, SslProtocols enabledSslProtocols, Boolean checkCertificateRevocation)
          /home/furt/github/wfurt-corefx/src/System.Net.Security/tests/FunctionalTests/ServerAllowNoEncryptionTest.cs(87,0): at System.Net.Security.Tests.ServerAllowNoEncryptionTest.ServerAllowNoEncryption_ClientNoEncryption_ConnectWithNoEncryption()
          --- End of stack trace from previous location where exception was thrown ---
          ----- Inner Stack Trace -----
          /home/furt/github/wfurt-corefx/src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs(91,0): at Interop.OpenSsl.AllocateSslContext(SslProtocols protocols, SafeX509Handle certHandle, SafeEvpPKeyHandle certKeyHandle, EncryptionPolicy policy, SslAuthenticationOptions sslAuthenticationOptions)
          /home/furt/github/wfurt-corefx/src/Common/src/System/Net/Security/Unix/SafeDeleteSslContext.cs(36,0): at System.Net.Security.SafeDeleteSslContext..ctor(SafeFreeSslCredentials credential, SslAuthenticationOptions sslAuthenticationOptions)
          /home/furt/github/wfurt-corefx/src/System.Net.Security/src/System/Net/Security/SslStreamPal.Unix.cs(111,0): at System.Net.Security.SslStreamPal.HandshakeInternal(SafeFreeCredentials credential, SafeDeleteContext& context, ArraySegment`1 inputBuffer, Byte[]& outputBuffer, SslAuthenticationOptions sslAuthenticationOptions)
      System.Net.Security.Tests.ServerAllowNoEncryptionTest.ServerAllowNoEncryption_ClientNoEncryption_ConnectWithNoEncryption [FINISHED] Time: 0.1360426s
```

Reply:
Root cause is this:
```
await sslStream.AuthenticateAsClientAsync("localhost", null, SslProtocolSupport.DefaultSslProtocols, false);
```
When I change DefaultSslProtocols (Tls, Tls11, Tls12, Tls13) to SslProtocols.Tls12 tests works as expected. That proves that OpenSSL 1.1.1 can do null encryption unless it negotiates Tls13

Would it be reasonable @bartonjs to update the test and assume that Tls12 is available on all platforms we care about? That seems better than disabling this test set like I did.

Reply:
> Would it be reasonable @bartonjs to update the test and assume that Tls12 is available on all platforms we care about?

Seems reasonable to me.  Leave a comment so that when it stops being reasonable we can point at this day and laugh.

(Ideally, by then, we'll just have dropped support for NoEncryption altogether, and just make AllowNoEncryption no-op.)